lib/webengine.cpp

```diff
@@ -17,24 +17,40 @@
 Q_GLOBAL_STATIC(SailfishOS::WebEngine, webEngineInstance)
 
 /*!
-  \class SailfishOS::WebEngine
-  \brief Provides access to the Web engine context
-  \inmodule SailfishWebView
+    \class SailfishOS::WebEngine
+    \brief Provides access to the Web engine context
+    \inmodule SailfishWebView
+    \inherits QMozContext
 
-  Singleton class which provides access to the Web engine context.
+    Singleton class which provides access to the Web engine context.
 */
 
 namespace SailfishOS {
 
 /*!
-  \brief Initialises the WebEngine class.
+    \brief Initialises the WebEngine class.
 
-  Initalises the WebEngine class. The \a profilePath sets the root folder for
-  the mozilla gecko profile. Once set the \a profilePath can't be changed.
+    Initalises the WebEngine class. The \a profilePath sets the root folder for
+    the mozilla gecko profile. Once set the \a profilePath can't be changed.
 
-  Multiple calls to initialize have no effect.
+    Multiple calls to initialize have no effect.
 
-  \sa initialize
+    This method will be called automatically during QML initialisation of the
+    WebView. However, a custom initialisation process can be applied as long as
+    it is called before the QML engine initialisation.
+
+    A custom initialization should perform the following steps.
+
+    1. Set up the environemnt, including setting \c{GRE_HOME} to the location
+       of the executing binary.
+    2. set up the profile path by calling \l setProfile
+    3. Add any needed component manifests by calling \l addComponentManifest (one
+       call for each manifest).
+
+    After which \l runEmbedding can be called to set the WebEngine event loop
+    running.
+
```

Comment:
While writing this part I realised that it's not currently really possible to override the default initialisation process after all, because there's no way to set the `isInitialized` value [here](https://github.com/sailfishos/sailfish-components-webview/blob/ebdacfbfa735d8b81233ce46d3a3cd7b33210b95/lib/webengine.cpp#L42), and so the built-initialisation will always happen as well.

Possible solutions to this:
1. Add a `WebEngine::overrideInitialization() { isInitialized = true; }` method that a developer can use to specify they don't want the standard initialisation to run.
2. Check whether `setProfile()` has been already set up and skip the initialisation if it has.
3. Make `WebEngine::initialize()` virtual so that devs can override it.
4. Since `Webengine` is really just an initialisation shim around `QMozContext`, document it in terms of reimplementing `WebEngine` as some other class that inherits `QMozContext`.
5. Don't support overriding the initialization process and don't include this documentation about it (probably in this case `setProfile()` and `runEmbedding()` needn't be included in the docs either??).

Any thoughts?

Reply:
WebEngine::initialized is currently called by QML plugin so for most QML app developers WebEngine is not needed. However, if you need to do custom steps I think documenting WebEngine::initialize would be good.

What if we'd alter initialize method a bit.
```
static void initialize(const QString &profilePath, bool runEmbedding = true);
```
If you need to delay calling the runEmbedding for a reason or another, you could now do it.

You really need those env variables, default components in order the WebView to function correctly. However, one may want to add more components. That said, I still think that more common case is that one needs some extra handling for the loaded content and then a frame script is enough.

Btw, I recently added example teardown of View + Engine to the QtMozEmbed test cases.

From application sandboxing point of view, I think it should be said that qstandard application paths are safe for profile paths as those include both application and organization name.  (FYI for @Tomin1 @okodron ).

Reply:
Thanks for the suggestion and helpful comments. I've made some changes, including adding the extra parameter and documenting it.

One thing that occurs is that in practice, because the call to `runEmbedding` in `initialize` is wrapped in a timer, in practice these are equivalent:

```
        webEngine::initialize(profilePath, false);
        webEngine->addComponentManifest(QLatin1String("/path/file.manifest"));
        QTimer::singleShot(0, webEngine, SLOT(runEmbedding()));
```
and
```
        webEngine::initialize(profilePath);
        webEngine->addComponentManifest(QLatin1String("/path/file.manifest"));
```
So we could maybe even get away without the extra parameter in this case. Nevertheless my personal feeling is that it's better to include it anyway, just in case there's a circumstance in which it's needed.